safeweb/http.go

```diff
@@ -0,0 +1,217 @@
+// - disallows form submissions to other origins
+// - disallows base URIs from other origins
+// - disallows mixed content when serving over HTTPS
+// - disallows embedding of resources from other origins
+const defaultCSP string = `
+	default-src 'self';
+	script-src 'self';
+	frame-ancestors 'none';
+	form-action 'self';
+	base-uri 'self';
+	block-all-mixed-content;
+	object-src 'none'
+`
+
+type Config struct {
+	// Hostname is the name of the tsnet service that will be created to host
+	// the application.
+	Hostname string
+
+	// Listener is a net.Listener that the server will host the application on.
+	// Providing a listening will override the use of tsnet and the Hostname field.
+	Listener net.Listener
+
+	// BrowserMux is the HTTP handler for any routes in your application
+	// that should be served to browsers. These requests will be subject to CSRF
+	// protection and will have browser-specific headers set in response.
+	BrowserMux *http.ServeMux
+
+	// APIMux is the HTTP handler for any routes in your application that
+	// should be served to non-browser clients.
+	APIMux *http.ServeMux
+
+	// PermittedCrossOriginHosts specifies the Access-Control-Allow-Origin header sent in response to pre-flight OPTIONS requests.
+	PermittedCrossOriginHosts []string
+	// PermittedCrossOriginMethods specifies the Access-Control-Allow-Methods header sent in response to pre-flight OPTIONS requests.
+	PermittedCrossOriginMethods []string
+
+	// CSRFSecret is the secret used to sign CSRF tokens. It must be 32 bytes long.
+	// TODO(@patrickod) do we want to keep this as a toggle? the intent is to
+	// prevent CSRF failures that occur due to the secret rotating between
+	// server restarts.
+	CSRFSecret []byte
+}
+
+func (c *Config) setDefaults() error {
+	if c.BrowserMux == nil {
+		c.BrowserMux = &http.ServeMux{}
+	}
+
+	if c.APIMux == nil {
+		c.APIMux = &http.ServeMux{}
+	}
+
+	if c.CSRFSecret == nil || len(c.CSRFSecret) == 0 {
+		c.CSRFSecret = make([]byte, 32)
+		if _, err := crand.Read(c.CSRFSecret); err != nil {
+			return fmt.Errorf("failed to generate CSRF secret: %w", err)
+		}
+	}
+
+	return nil
+}
+
+type Server struct {
+	Config
+	h *http.Server
+}
+
+func wrapHandler(config Config) func(*http.ServeMux, *http.ServeMux) http.Handler {
+	// only serve HTTPS over tsnet when a listener is not provided.
+	serveHTTPS := config.Listener == nil
+	csrfProtect := csrf.Protect(config.CSRFSecret, csrf.Secure(serveHTTPS))
+
+	return func(browserHandler, apiHandler *http.ServeMux) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// determine which of the two muxes should serve the request
+			var h http.Handler
+			var apiRequest bool
+			h, p := browserHandler.Handler(r)
+			if p == "" {
+				h = apiHandler
+				apiRequest = true
+			}
+
+			if apiRequest {
+				switch r.Method {
+				// disallow `application/x-www-form-urlencoded` content-type for all API requests
+				case "POST", "PUT", "PATCH":
```

Comment:
I think we should drop this filter. There should be no API request with this content-type ever.